pkg/util/file_indexer_test.go

```diff
@@ -0,0 +1,129 @@
+package util
+
+import (
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/openshift/odo/pkg/testingutil/filesystem"
+)
+
+func TestCheckGitIgnoreFile(t *testing.T) {
+
+	// create a fake fs in memory
+	fs := filesystem.NewFakeFs()
+	// create a context directory on fake fs
+	contextDir, err := fs.TempDir(os.TempDir(), "context")
+	if err != nil {
+		t.Error(err)
+	}
+
+	gitignorePath := filepath.Join(contextDir, ".gitignore")
+
+	tests := []struct {
+		testName        string
+		create          bool
+		gitIgnoreCreate func(create bool, contextDir string, fs filesystem.Filesystem) error
+		directory       string
+		want            string
+		wantErr         bool
+	}{
+		{
+			testName:        "Test when .gitignore does not exist",
+			create:          true,
+			gitIgnoreCreate: mockDirectoryInfo,
+			directory:       contextDir,
+			want:            gitignorePath,
+			wantErr:         false,
+		},
+		{
+			testName:        "Test when .gitignore exists",
+			create:          false,
+			gitIgnoreCreate: mockDirectoryInfo,
+			directory:       contextDir,
+			want:            gitignorePath,
+			wantErr:         false,
+		},
+	}
+
+	for _, tt := range tests {
+
+		err := tt.gitIgnoreCreate(tt.create, tt.directory, fs)
+		if err != nil {
+			t.Error(err)
+		}
+
+		t.Run(tt.testName, func(t *testing.T) {
+
+			gitIgnoreFilePath, err := checkGitIgnoreFile(tt.directory, fs)
+
+			if tt.want != gitIgnoreFilePath {
+				t.Errorf("checkGitIgnoreFile unexpected error %v, while creating .gitignore file", err)
+			}
+
+			if !tt.wantErr == (err != nil) {
+				t.Errorf("checkGitIgnoreFile unexpected error %v, wantErr %v", err, tt.wantErr)
+			}
+		})
+	}
+
+}
+
+func TestAddOdoFileIndex(t *testing.T) {
+
+	// create a fake fs in memory
+	fs := filesystem.NewFakeFs()
+	// create a context directory on fake fs
+	contextDir, err := fs.TempDir(os.TempDir(), "context")
+	if err != nil {
+		t.Error(err)
+	}
+
+	gitignorePath := filepath.Join(contextDir, ".gitignore")
+
+	tests := []struct {
+		testName        string
+		create          bool
+		gitIgnoreCreate func(create bool, contextDir string, fs filesystem.Filesystem) error
+		directory       string
+		wantErr         bool
+	}{
+		{
+			testName:        "Test when odo-file-index.json added to .gitignore",
+			create:          false,
+			gitIgnoreCreate: mockDirectoryInfo,
+			directory:       gitignorePath,
+			wantErr:         false,
+		},
+	}
+
+	for _, tt := range tests {
+
+		err := tt.gitIgnoreCreate(tt.create, tt.directory, fs)
+		if err != nil {
+			t.Error(err)
+		}
+
+		t.Run(tt.testName, func(t *testing.T) {
+
+			err := addOdoFileIndex(tt.directory, fs)
+
+			if !tt.wantErr == (err != nil) {
+				t.Errorf("addOdoFileIndex unexpected error %v, wantErr %v", err, tt.wantErr)
+			}
+
+		})
+	}
+}
+
+func mockDirectoryInfo(create bool, contextDir string, fs filesystem.Filesystem) error {
+
+	if !create {
```

Comment:
I'm a little confused. This condition will be satisfied when `create` is set to `false`. So when it's `false`, we're creating the file/directory and when it's `true`, we're not doing anything of the sort. This is by design, right?

Just trying to affirm my understanding. Correct me if I'm wrong.

Reply:
Correct @dharmit 

In the test case where .gitignore directory does not exist, create will be set to true as we don't want that directory to exist.

Reply:
It feels counter-intuitive to me. In the scenario that you explained, why not have `create` set to `false` since we already have what we need? Since you have used the word "directory" twice, aren't we talking about `.gitignore` file here? I'm feeling lost here. :disappointed:

Reply:
"directory" is the kind of context directory you can say where the component will get created. That directory we need to pass to the functions. 

Again in second unit test https://github.com/openshift/odo/pull/2441/files#diff-0292f6b604e75b79cef052987afa7fb5R72 we should be having .gitignore file as already checked from https://github.com/openshift/odo/pull/2441/files#diff-0292f6b604e75b79cef052987afa7fb5R11

Reply:
In the earlier comment, you wrote:
> In the test case where .gitignore directory does not exist

so I thought you're talking about a directory called `.gitinore`. Hence the question.

I'm still confused as to why we're creating a directory when `create` bool is `false`. :confused:

Reply:
> In the earlier comment, you wrote:
> 
> > In the test case where .gitignore directory does not exist
> 
> so I thought you're talking about a directory called `.gitinore`. Hence the question.
> 
> I'm still confused as to why we're creating a directory when `create` bool is `false`. 😕

Because whatever false directory we are creating will get created in RAM and once that unit test is complete it will delete the directory and everything.

As for https://github.com/openshift/odo/pull/2441/files#diff-0292f6b604e75b79cef052987afa7fb5R72 this unit test .gitignore file should be there to add odo-file-index into .gitignore.

Hence ```create``` bool is set to false, so that we can have .gitignore file there in the directory.

Reply:
> so I thought you're talking about a directory called `.gitinore`. Hence the question.
There are two different unit test, one for checking ```.gitignore``` exists or not. If not then create one. Second is to add ```odo-file-index.json``` into ```.gitignore```.

If we can link these two unit test then ```create``` bool will be set to ```true```. So is there any way we can do that as I am not aware of that. Please do suggest if there is any to return ```.gitignore``` path from the first unit test I mentioned above. If not then we have have to create ```.gitignore``` directory as we are using fake filesystem with ```afero```.